src/Npgsql/TypeHandlers/CompositeHandlers/CompositeHandler.cs

```diff
@@ -90,25 +97,89 @@ public override int ValidateAndGetLength(T value, ref NpgsqlLengthCache? lengthC
-            var length = sizeof(int) + sizeof(int) * 2 * _memberHandlers.Length;
-            foreach (var member in _memberHandlers)
+            var memberHandlers = _memberHandlers!;
+            var length = sizeof(int) + sizeof(int) * 2 * memberHandlers.Length;
+            foreach (var member in memberHandlers)
                 length += member.ValidateAndGetLength(value, ref lengthCache);
 
             return lengthCache.Lengths[position] = length;
         }
 
-        CompositeMemberHandler<T>[] CreateMemberHandlers() => CreateMemberHandlers((PostgresCompositeType)PostgresType, _typeMapper, _nameTranslator);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        void Initialize()
+        {
+            if (_memberHandlers is null)
+                InitializeCore();
+        }
 
-        static CompositeMemberHandler<T>[] CreateMemberHandlers(PostgresCompositeType pgType, ConnectorTypeMapper typeMapper, INpgsqlNameTranslator nameTranslator)
+        void InitializeCore()
         {
-            var clrType = typeof(T);
+            var pgType = (PostgresCompositeType)PostgresType;
             var pgFields = pgType.Fields;
-            var clrHandlers = new CompositeMemberHandler<T>[pgFields.Count];
-            var clrHandlerCount = 0;
 
-            var clrHandlerType = IsValueType<T>.Value
+            var clrType = typeof(T);
+            var clrMemberHandlers = new CompositeMemberHandler<T>[pgFields.Count];
+            var clrMemberHandlerCount = 0;
+            var clrMemberHandlerType = IsValueType<T>.Value
                 ? typeof(CompositeStructMemberHandler<,>)
                 : typeof(CompositeClassMemberHandler<,>);
+            var clrConstructorHandlerType = typeof(CompositeParameterHandler<>);
+
+            ConstructorInfo? clrDefaultConstructor = null;
+            CompositeConstructorHandler<T>? clrConstructorHandler = null;
+
+            foreach (var clrConstructor in clrType.GetConstructors())
+            {
+                var clrParameters = clrConstructor.GetParameters();
+                if (clrParameters.Length != pgFields.Count)
+                {
+                    if (clrParameters.Length == 0)
+                        clrDefaultConstructor = clrConstructor;
+
+                    continue;
+                }
+
+                var clrParameterHandlers = new CompositeParameterHandler[pgFields.Count];
+                var clrParameterHandlerCount = 0;
+
+                foreach (var clrParameter in clrParameters)
+                {
+                    var attr = clrParameter.GetCustomAttribute<PgNameAttribute>();
+                    var name = attr?.PgName ?? _nameTranslator.TranslateMemberName(clrParameter.Name!);
+
+                    for (var pgFieldIndex = pgFields.Count - 1; pgFieldIndex >= 0; --pgFieldIndex)
+                    {
+                        var pgField = pgFields[pgFieldIndex];
+                        if (pgField.Name != name)
```

Comment:
What happens if the name matches, but not the type? We could verify that the parameter's type is mapped to a handler for the corresponding member's PG type. 

But I admit this is a bit contrived.

Reply:
The code will fail while reading/writing. There is no way to get a handler by OID and check that it can handle the CLR type. Keeping this in my mind.